### PR TITLE
Simplify the cache service API and store all state in a database

### PR DESCRIPTION
**Pull request by kraih**
This cache service PR is a bit more risky than my previous ones, so should not be applied without acceptance testing or while we need OSD to be 100% reliable. It removes pretty much everything that's not absolutely necessary from the API.

The cache service had this rather confusing feature where it would store the current job state in two places, first in memory, and then in SQLite (while the actual background job was in SQLite from the start). That means if you restarted the cache service at the wrong time you would get strange race conditions. So this PR moves all job state into SQLite to make it persistent across restarts, and the concept of session tokens is gone completely (i could not actually figure out what it was meant to achieve, possibly cache invalidation...but there were no tests proving that theory). The Minion background jobs also had to communicate with the REST API while processing jobs, that IPC layer is also no longer needed.

In fact the cache service API now only has 3 routes remaining, and the client 3 corresponding methods with the same name (`/enqueue` to tell the cache service what to download, `/info` to get general information about the cache service and `/status` to check the state of a download and get meta data).

So the code and the download life cycle are much easier to follow now, which should help a lot with debugging.

### CODE REVIEW

**Comment by kraih** [comment]
I'll deploy this on the staging machines for acceptance testing.

**Comment by kraih** [comment]
Almost forgot to mention, this PR also reduces the number of HTTP requests required to communicate with the cache service API significantly. Download state, result and output are all provided by the same API endpoint (`/status`), but in the previous version of the client the same data structure had to be requested three times.

**Comment by kraih** [comment]
Maybe it would help potential reviewers to know that `isotovideo.pm` is the only place where the worker actually communicates with the cache service (and only through `OpenQA::CacheService::Client`). Otherwise the cache service is completely self contained now.

File: lib/OpenQA/CacheService/Response.pm [hunk 1]
```diff
@@ -1,4 +1,4 @@
-# Copyright (C) 2019 SUSE LLC
+# Copyright (C) 2018 SUSE LLC
```
Comment (Martchus) [inline lib/OpenQA/CacheService/Response.pm]: Git makes it look like this has been changed from 2019 to 2018. Maybe you add both years.

File: t/25-cache-service.t [hunk 2]
```diff
@@ -144,16 +144,11 @@ sub test_download {
...
 
-    my $state = $cache_client->status($asset_request);
-    $state = $cache_client->status($asset_request) until ($state ne OpenQA::CacheService::Model::Cache::STATUS_IGNORE);
-
-    # After IGNORE, only DOWNLOAD status could follow, but it could be fast enough to not catch it
-    $state = $cache_client->status($asset_request);
-    $state = $cache_client->status($asset_request)
-      until ($state ne OpenQA::CacheService::Model::Cache::STATUS_DOWNLOADING);
+    my $status = $cache_client->status($asset_request);
+    $status = $cache_client->status($asset_request) until !$status->is_downloading;
```
Comment (Martchus) [inline t/25-cache-service.t]: Not even a small sleep between the polls?

Comment (kraih) [reply t/25-cache-service.t]: Yea, that surprised me too. The test jobs are very fast though, so it probably doesn't really matter.

**Comment by kraih** [comment]
Tests on the staging machines showed some issues, i will have to make some adjustments.

**Comment by okurz** [comment]
Great. Good that we have the staging tests :)

File: lib/OpenQA/CacheService/Controller/API.pm [hunk 3]
```diff
@@ -32,62 +24,36 @@ sub info {
...
+      unless my $job_info = $job->info;
 
-        my $job_info = $job->info;
-        return $self->render(json => {error => 'Job info not available.'}, status => 404) unless $job_info;
-        $res{result} = $job_info->{result};
-        $res{output} = $job_info->{notes}->{output};
-    }
+    # The job will finish early if another job is already downloading, so we have to wait for the lock too
+    $self->render(json => {status => 'processed', result => $job_info->{result}, output => $job_info->{notes}{output}})
+      if $job_info->{state} eq 'finished' && !$self->progress->is_downloading($job_info->{notes}{lock});
```
Comment (kalikiana) [inline lib/OpenQA/CacheService/Controller/API.pm]: I'm unclear... how is this `render` call waiting for the lock? I might be misunderstanding the comment, or maybe I'm not clear what the code is meant to do.

**Comment by kraih** [comment]
The issue wasn't so bad after all, and this PR should be fine now. But i'll run a few more tests on Monday.

File: lib/OpenQA/CacheService/Controller/API.pm (hunk 3 above)
Comment (kraih) [reply lib/OpenQA/CacheService/Controller/API.pm]: The comment is about the `if $job_info->{state} eq 'finished' && !$self->progress->is_downloading($job_info->{notes}{lock})` check. If there is already another background job downloading the asset, our job will finish immediately. And then we have to wait for the lock to be freed by the other background job (by calling `->progress->is_downloading`). This action (`/status`) gets called over and over until the status in the JSON response is `processed`.

Comment (kraih) [reply lib/OpenQA/CacheService/Controller/API.pm]: Feel free to suggest a better explanation, i'd like for this to be as simple to understand as possible. That's why i'm making all these changes in the first place. 😉

**Comment by kraih** [comment]
I should probably mention that the latest commit made the lock logic a lot less smart. Previously the lock could even be active while there was no Minion job running. Just to save a few CPU cycles at the cost of code complexity. That was a bad idea, and i have now switched to the standard Minion pattern of using locks only while the Minion job is active.

File: lib/OpenQA/CacheService/Controller/API.pm (hunk 3 above)
Comment (kalikiana) [reply lib/OpenQA/CacheService/Controller/API.pm]: Right. Thanks for clarifying.

I think it's the use of "we" and "wait" here that makes me think of blocking on the result which isn't really what's happening and the status call isn't doing anything in the first place.

The job will wait on the lock. And then the real consumer will take it from there.

Comment (kalikiana) [reply lib/OpenQA/CacheService/Controller/API.pm]: It's very possible that I'm getting lost in semantics. But from what I see in the code, the status call is not much more than an API doing no actual work.

Comment (kraih) [reply lib/OpenQA/CacheService/Controller/API.pm]: I'll turn this route into a `GET /status/<id>`, that should make its nature more obvious.

**Comment by okurz** [comment]
my approval still holds however the full stack test fails now and as your work is related to caching I guess the test could show a real issue

**Comment by kraih** [comment]
@okurz It is a real issue, but more complicated than it looks. The test passes locally and only fails in Circle CI.

File: lib/OpenQA/CacheService/Controller/API.pm (hunk 3 above)
Comment (kalikiana) [reply lib/OpenQA/CacheService/Controller/API.pm]: Even better :+1:

**Comment by coolo** [comment]
We found the job 5 is set to incomplete as it can't reach the webui 'not to be considered dead' - even before creating the symlink.

How this PR affects this behaviour is unclear, but at least we got closer to the problem.

**Comment by kraih** [comment]
We have found the problem, it was a bug in `full-stack.t`. This PR just managed to trigger it consistently due to some small timing changes. The fix is included as the final commit, so this PR should be ready for merge now.

**Pull request merged**